unicon_runner/executor/variants/sandbox/executor.py

```diff
@@ -32,20 +38,37 @@ async def _execute(self, request: Request, request_id: str, folder_path: str) ->
             pyproject_file = self.pyproject_template.render()
             f.write(pyproject_file)
 
+        with open(os.path.join(code_folder_path, "__init__.py"), "w") as f:
+            f.write("")
+
+        with open(os.path.join(folder_path, "requirements.txt"), "w") as f:
+            if "requirements" in request.environment.options:
+                f.write(request.environment.options["requirements"])
+
         # 2. Cd into temp folder and run uv sync && uv run entry
         proc = await asyncio.create_subprocess_shell(
-            f"SANDBOX=1 SANDBOX_LEVEL=1 QUIET_MODE=1 {self.CONTY} "
-            f"--bind {os.path.abspath(folder_path)} ~/{folder_path} "
-            f"--ro-bind {os.path.abspath(self.RUN_SCRIPT)} ~/{self.RUN_SCRIPT} "
-            # NOTE: `uv` binary is assumed to be stored under `~/.cargo/bin/`
-            # We are using `uv` as the environment manager and program runner
-            f"--ro-bind ~/.cargo ~/.cargo "
-            f"./{self.RUN_SCRIPT} {folder_path} {self.CODE_FOLDER_NAME}/{request.entrypoint} {request.environment.memory_limit * 1024} {request.environment.time_limit}",
+            f"UV_CONCURRENT_INSTALLS=1 {self.INSTALL_SCRIPT} {folder_path}",
             stdout=asyncio.subprocess.PIPE,
             stderr=asyncio.subprocess.PIPE,
         )
+        await proc.wait()
+
+        async with self.lock:
+            proc = await asyncio.create_subprocess_shell(
+                f"SANDBOX=1 SANDBOX_LEVEL=1 QUIET_MODE=1 UV_CONCURRENT_INSTALLS=1 {self.CONTY} "
+                # proc = await asyncio.create_subprocess_shell(
+                #     f"UV_CONCURRENT_INSTALLS=1 {self.INSTALL_SCRIPT} {folder_path} && SANDBOX=1 SANDBOX_LEVEL=1 QUIET_MODE=1 UV_CONCURRENT_INSTALLS=1 {self.CONTY} "
+                f"--bind {os.path.abspath(folder_path)} {folder_path} "
+                f"--ro-bind {os.path.abspath(self.RUN_SCRIPT)} ~/{self.RUN_SCRIPT} "
+                # NOTE: `uv` binary is assumed to be stored under `~/.cargo/bin/`
+                # We are using `uv` as the environment manager and program runner
+                f"--ro-bind ~/.cargo ~/.cargo "
+                f"./{self.RUN_SCRIPT} {folder_path} {self.CODE_FOLDER_NAME}/{request.entrypoint} {request.environment.memory_limit * 1024} {request.environment.time_limit}",
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
```

Comment:
now that program assembly is done by `unicon-backend`, we wouldnt need to handle multiple test-cases since programs are gonna be sent to `unicon-runner` at a testcase level

Reply:
<will handle in next pr>